### PR TITLE
fix: produce tombstone records

### PR DESCRIPTION
**Pull request by fasazevedo**
# Description

Adds support for tombstone records production.

Fixes https://github.com/Farfetch/kafkaflow/issues/516

## How Has This Been Tested?

This was tested via the provided unit and integration tests.

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have added tests to cover my changes
-   [x] I have made corresponding changes to the documentation

### Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

**Review by brmagadutra — APPROVED** [review]

File: src/KafkaFlow/Middlewares/Serializer/SerializerProducerMiddleware.cs [hunk 1]
```diff
@@ -38,18 +38,21 @@ public async Task Invoke(IMessageContext context, MiddlewareDelegate next)
     {
         await _typeResolver.OnProduceAsync(context);
 
-        byte[] messageValue;
+        byte[] messageValue = null;
```
Comment (esskar) [inline src/KafkaFlow/Middlewares/Serializer/SerializerProducerMiddleware.cs:41]: Is null the safest here? What about 

`byte[] messageValue = Array.Empty<byte>();`?

**Review by JoaoRodriguesGithub — APPROVED** [review]

**Review by kikofps — APPROVED** [review]

File: src/KafkaFlow/Middlewares/Serializer/SerializerProducerMiddleware.cs (hunk 1 above)
Comment (fasazevedo) [reply src/KafkaFlow/Middlewares/Serializer/SerializerProducerMiddleware.cs:41]: Hello esskar!

I may be understanding this wrong, but I believe we want to produce record entries with null message values (the so called tombstone records), so in that case it's better consider it as null? What do you think?

**Review by ruiqbarbosa — APPROVED** [review]

File: src/KafkaFlow/Middlewares/Serializer/SerializerProducerMiddleware.cs (hunk 1 above)
Comment (fasazevedo) [reply src/KafkaFlow/Middlewares/Serializer/SerializerProducerMiddleware.cs]: Hello again esskar! 

In spite of the fact that null initialization was ok in our tests, we decided that your suggestion was actually more bullet proof. I've changed the code to apply that and so I'll close this now. 

Thank you! 🙇

**Review by ailtonguitar — APPROVED** [review]

**Comment by mrt181** [comment]
How is the fix supposed to work?

Commit [082baf3](https://github.com/Farfetch/kafkaflow/commit/082baf349c03f513bcfeb95979efbfd615e86546)

adds this
```diff

-        byte[] messageValue;
+        byte[] messageValue = Array.Empty<byte>();

-       using (var buffer = s_memoryStreamManager.GetStream())
+        if (context.Message.Value is not null)
-        {
-            await _serializer
-                .SerializeAsync(
-                    context.Message.Value,
-                    buffer,
-                    new SerializerContext(context.ProducerContext.Topic))
-                .ConfigureAwait(false);
-
-            messageValue = buffer.ToArray();
+            using (var buffer = s_memoryStreamManager.GetStream())
+            {
+                await _serializer
+                    .SerializeAsync(
+                        context.Message.Value,
+                        buffer,
+                        new SerializerContext(context.ProducerContext.Topic))
+                    .ConfigureAwait(false);
+
+                messageValue = buffer.ToArray();
+            }


The [SerializerProducerMiddleware](https://github.com/Farfetch/kafkaflow/commit/082baf349c03f513bcfeb95979efbfd615e86546#diff-00dcad3a637cfe5d6ad0ec93d5c61d5685a6a188ad69f21a22d1f1c7c7239838) sets the value to an empty byte array and if the value is not null calls any provided serializers, otherwise it sends the emoty byte array.

So even if you had a custom one that handles null it would never be called.

It even updates the docs [serializer-middleware.md](https://github.com/Farfetch/kafkaflow/commit/082baf349c03f513bcfeb95979efbfd615e86546#diff-5698cd31e603260c9e44f106e2cc0c44da52b958338cd91c9240dd40504ec696) to say that to create a tombstone you must send a null value not an empty byte array?!

**Pull request merged**